### PR TITLE
sandy | fire

### PR DESCRIPTION
**Pull request by crypt0kitty**
# Heaps Practice

Congratulations! You're submitting your assignment!

## Comprehension Questions

Question | Answer
:------------- | :-------------
How is a Heap different from a Binary Search Tree? | In a heap each parent has a specific order relationship with its children. With a Min-Heap every parent node is less than its child nodes. In a Max-Heap, every parent node is greater than its child nodes.
Could you build a heap with linked nodes? | Yes, but an array is preferred because you can use the formula to get the children of a node: left_child = i * 2 + 1 right_child = i * 2 + 2
Why is adding a node to a heap an O(log n) operation? | Worst case it will traverse from bottom to the root node
Were the `heap_up` & `heap_down` methods useful?  Why? | It helped keep things organized/in order

### CODE REVIEW

File: lib/min_heap.rb [hunk 1]
```diff
@@ -13,19 +13,31 @@ def initialize
...
+  def last_index
+    @store.length - 1
+  end
+
   # This method adds a HeapNode instance to the heap
-  # Time Complexity: ?
-  # Space Complexity: ?
+  # Time Complexity: O(log n). Pushing O(1) && heap_up O(log n)
+  # Space Complexity: O(log n) 
   def add(key, value = key)
```
Comment (CheezItMan) [inline lib/min_heap.rb:23]: 👍

File: lib/min_heap.rb [hunk 2]
```diff
@@ -13,19 +13,31 @@ def initialize
...
   end
 
   # This method removes and returns an element from the heap
-  #   maintaining the heap structure
-  # Time Complexity: ?
-  # Space Complexity: ?
+  # maintaining the heap structure
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n)
   def remove()
```
Comment (CheezItMan) [inline lib/min_heap.rb:33]: 👍

File: lib/min_heap.rb [hunk 3]
```diff
@@ -44,28 +56,55 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(1)
+  # Space complexity: O(1)
   def empty?
```
Comment (CheezItMan) [inline lib/min_heap.rb:61]: 👍

File: lib/min_heap.rb [hunk 4]
```diff
@@ -44,28 +56,55 @@ def to_s
...
   private
-
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(log n)
+  # Space complexity: O(log n)
   def heap_up(index)
```
Comment (CheezItMan) [inline lib/min_heap.rb:71]: 👍 Very readable and elegant code.

File: lib/min_heap.rb [hunk 5]
```diff
@@ -44,28 +56,55 @@ def to_s
...
 
   # This helper method takes an index and 
-  #  moves it up the heap if it's smaller
-  #  than it's parent node.
+  # moves it up the heap if it's smaller
+  # than it's parent node.
+  # Time complexity: O(log n)
+  # Time complexity: O(log n)
+
   def heap_down(index)
```
Comment (CheezItMan) [inline lib/min_heap.rb:91]: 👍

**Review by CheezItMan — COMMENTED** [review]
The heap class works and works well.  You don't have the heapsort, but otherwise this is good.

**Comment by crypt0kitty** [comment]
> The heap class works and works well. You don't have the heapsort, but otherwise this is good.

Oh no, I forgot to add that part.

**Pull request closed without merging**